### ISSUE TITLE
(Beta 5) "No MaterialLocalizations found" Error when using custom wiredash localizations delegate

### ISSUE THREAD

**Issue by mdrideout** [wiredashio/wiredash-sdk#223]
It appears that when I activate Wiredash using a localizationDelegate in the WiredashOptionsData to override the UI strings, the localizations are stripped from the rest of the app.

The Wiredash translations are working fine.

I get the following error attempting to open Wiredash. The app runs fine until Wiredash is attempted to be opened. (The AppBar uses `AppLocalizations.of(context)!.translationString`). Additional similar errors for each line using AppLocalizations for the given context.

```
======== Exception caught by widgets library =======================================================
The following assertion was thrown building AppBar(dirty, dependencies: [_ScrollNotificationObserverScope, _LocalizationsScope-[GlobalKey#9becd]], state: _AppBarState#5aac7):
No MaterialLocalizations found.

AppBar widgets require MaterialLocalizations to be provided by a Localizations widget ancestor.
The material library uses Localizations to generate messages, labels, and abbreviations.

To introduce a MaterialLocalizations, either use a MaterialApp at the root of your application to include them automatically, or add a Localization widget with a MaterialLocalizations delegate.
```

<img width="847" alt="wiredash-error" src="https://user-images.githubusercontent.com/3958274/178590823-4c3c5f77-1451-47a4-98e0-6f2c048c19f8.png">

**main.dart**: Using MaterialApp.router with normal [AppLocalizations implementation from flutter docs](https://docs.flutter.dev/development/accessibility-and-localization/internationalization). 
```dart
          child: MaterialApp.router(
            locale: _locale,
            supportedLocales: supportedLocales,
            localizationsDelegates: AppLocalizations.localizationsDelegates,
            debugShowCheckedModeBanner: false,
            scaffoldMessengerKey: rootScaffoldMessengerKey,
            title: 'XXX',
            theme: ConnectTheme.themeData(context),
            routerDelegate: RoutemasterDelegate(
              observers: [
                ConnectRouteObserver(
                  firebaseAnalytics: firebaseAnalytics,
                  user: user,
                  ref: ref,
                ),
              ],
              routesBuilder: (context) {
                // Watch for user state changes to rebuild our RouteMap based on user login state
                RouteMap _routeMap = ref.watch(routeMapProvider);
                return _routeMap;
              },
            ),
            routeInformationParser: const RoutemasterParser(),
          ),
```

**Wrapper widget for all screens, where Wiredash is parent**
```dart
              // User exists (Wiredash must be here because we need access to a BuildContext after Localizations have been initialized.
              // in order to get translations for our labels.
              // and this widget is a wrapper for all authenticated user screens.
              return Wiredash(
                projectId: wiredashCredentials.projectId,
                secret: wiredashCredentials.secret,
                options: WiredashOptionsData(
                  // Locale in wiredash does not support defining a country code as of 1.0.0-beta5, so derive from just the language code
                  locale: Locale.fromSubtags(languageCode: _locale.languageCode),
                  localizationDelegate: const WiredashCustomLocalizationsDelegate(),
                ),
                feedbackOptions: WiredashFeedbackOptions(
                  askForUserEmail: true,
                  labels: getWiredashLabels(context),
                ),
                child: screen,
              );
```

**Wiredash activation (via drawer)**
```dart
                    // Set Wiredash Data
                    String? userEmail = _user?.email ?? "email not available";
                    Wiredash.of(context).setUserProperties(userEmail: userEmail);
                    Wiredash.of(context).setBuildProperties(buildNumber: _buildNumber, buildVersion: _version);

                    // Launch Wiredash Interface
                    Wiredash.of(context).show(inheritMaterialTheme: true);
```

**Wiredash custom localizations delegate**
```dart
import 'package:connect/models/locale/locale_providers.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:wiredash/wiredash.dart';

/// Wiredash Custom Localizations Delegate
/// Use this to provide your own texts that match the style of your app.
/// Part of WiredashOptionsData parameters.
/// See Example: https://github.com/wiredashio/wiredash-sdk/blob/stable/example/lib/custom_demo_translations.dart
class WiredashCustomLocalizationsDelegate extends LocalizationsDelegate<WiredashLocalizations> {
  const WiredashCustomLocalizationsDelegate();

  @override
  bool isSupported(Locale locale) {
    // Check if the provided locale language code exists in any of our supportedLocale language codes
    return supportedLocales.any((supportedLocale) => supportedLocale.languageCode == locale.languageCode);
  }

  @override
  Future<WiredashLocalizations> load(Locale locale) {
    switch (locale.languageCode) {
      case "en":
        {
          return SynchronousFuture(_CustomTranslationsEn());
        }

      case "de":
        {
          return SynchronousFuture(_CustomTranslationsDe());
        }

      default:
        return SynchronousFuture(_CustomTranslationsEn());
    }
  }

  @override
  bool shouldReload(covariant LocalizationsDelegate<WiredashLocalizations> old) => false;
}

/// Override English Wiredash Strings
class _CustomTranslationsEn extends WiredashLocalizationsEn {
  @override
  String get feedbackStep1MessageDescription =>
      "Add description for your request functions and problems you encountered in xxx";

  @override
  String get feedbackStep1MessageHint =>
      "I want to order ink from xxx... / There's an unknown error when I check printer monitoring...";
}

/// Override German Wiredash Strings
class _CustomTranslationsDe extends WiredashLocalizationsEn {
  @override
  String get feedbackStep1MessageDescription =>
      "Fügen Sie eine Beschreibung der von Ihnen angefragten Funktionen und der Probleme hinzu, auf die Sie in xxx gestoßen sind";

  @override
  String get feedbackStep1MessageHint =>
      "Ich möchte Tinte bei xxx bestellen... / Es gibt einen unbekannten Fehler, wenn ich die Druckerüberwachung prüfe...";
}
```

**Comment by passsy** [comment]
Thanks for such detailed reporting 💙

The issue is caused because `Wiredash` sits below `MaterialApp` that internally injects the `Localizations` widget into the tree. Wiredash indeed overrides the `Localizations` widget, dropping existing delegates.

The correct order is to have `Wiredash` sit above the `MaterialApp`. From the code you provided that should be doable because `_locale` is known when you construct `MaterialApp`.

```dart
return Wiredash(
  projectId: wiredashCredentials.projectId,
  secret: wiredashCredentials.secret,
  options: WiredashOptionsData(
    // Locale in wiredash does not support defining a country code as of 1.0.0-beta5, so derive from just the language code
    locale: Locale.fromSubtags(languageCode: _locale.languageCode),
    localizationDelegate: const WiredashCustomLocalizationsDelegate(),
  ),
  feedbackOptions: WiredashFeedbackOptions(
    askForUserEmail: true,
    labels: getWiredashLabels(context),
  ),
  child: MaterialApp.router(
    locale: _locale,
    supportedLocales: supportedLocales,
    localizationsDelegates: AppLocalizations.localizationsDelegates,
    debugShowCheckedModeBanner: false,
    scaffoldMessengerKey: rootScaffoldMessengerKey,
    title: 'XXX',
    theme: ConnectTheme.themeData(context),
    routerDelegate: RoutemasterDelegate(
      observers: [
        ConnectRouteObserver(
          firebaseAnalytics: firebaseAnalytics,
          user: user,
          ref: ref,
        ),
      ],
      routesBuilder: (context) {
        // Watch for user state changes to rebuild our RouteMap based on user login state
        RouteMap _routeMap = ref.watch(routeMapProvider);
        return _routeMap;
      },
    ),
    routeInformationParser: const RoutemasterParser(),
  );
```
---

Nevertheless, I look to find a way to continue providing existing localizationDelegates. Should be doable.

**Comment by mdrideout** [comment]
The issue I had with putting Wiredash above `MaterialApp.router` is that in order to get the translations for our labels (under feedbackOptions), we need access to a BuildContext that has AppLocalizations instantiated, which only happens below MaterialApp.

Example of our translated label list:

```dart
/// Wiredash Labels List
List<Label> getWiredashLabels(BuildContext context) {
  return <Label>[
    Label(id: "label-vku...", title: AppLocalizations.of(context)!.labelsPraise),
    Label(id: "label-oav...", title: AppLocalizations.of(context)!.labelsBug),
    Label(id: "label-da6...", title: AppLocalizations.of(context)!.labelsImprovement),
    Label(id: "label-acn...", title: AppLocalizations.of(context)!.labelsNewIdea),
  ];
}
```

Throws error:
```
The following _CastError was thrown building Consumer(dirty, dependencies: [UncontrolledProviderScope], state: _ConsumerState#2a30a):
Null check operator used on a null value

The relevant error-causing widget was: 
  Consumer Consumer:file:///Users/matt/repos/roland/rdgconnect/lib/main.dart:150:12
When the exception was thrown, this was the stack: 
#0      getWiredashLabels (package:connect/models/wiredash/constants_wiredash.dart:18:70)
```

Can you suggest a better option for getting translated labels from our .arb files if Wiredash needs to be above MaterialApp?

```dart
  @override
  String get feedbackStep1MessageHint =>
      "Ich möchte Tinte bei xxx bestellen... / Es gibt einen unbekannten Fehler, wenn ich die Druckerüberwachung prüfe...";
```

For the overrides of the Wiredash Localizations, we had to use strings in the getters where we would really prefer to use the .arb translation files we are already using for the rest of the app. Is this possible? We don't seem to have a build context here to get the translations either. And the alternative method of using a delegate loader like this requires an async function to load, which is not a valid override.

```dart
  @override
  String get feedbackStep1MessageDescription async {
    final l10n = await AppLocalizations.delegate.load(locale);
    return l10n.labelsNewIdea
  }
```

Our locale is dynamic using Riverpod state, so when the user changes locale, the state of locale updates rebuilding the widgets. So when the locale is changed, we also do not have an opportunity to introduce an async function during that rebuild.

```dart
@override
  Widget build(BuildContext context) {
    return Consumer(builder: (context, ref, child) {
      // Watch for locale changes
      Locale _locale = ref.watch(localeProvider);
...
    return Wiredash(
      ...
      child: MaterialApp.router(
        ...
```

**Comment by passsy** [comment]
Got it. Yes, the label localization needs to be revisited and smoothed out. I'll write an example for this.

Meanwhile, please try the fix in #224 I just merged. It allows you to place `Wiredash` below `MaterialApp`

```yaml
dependencies:
  wiredash:
    git:
      url: https://github.com/wiredashio/wiredash-sdk.git
      revision: a4c3a913eca1ce8c63f10bdcd9ebf207fbfa2f54
```

---

The best position for Wiredash is before any UI and after all non-UI things that need to be provided. i.e.

```
--- Non UI ---
FeatureFlags
Localizations
NetworkingProvider
...
Wiredash
--- UI ---
MaterialApp
Scaffold
...
```

**Comment by mdrideout** [comment]
Thank you so much I will try that out.

I edited the response, so not sure if you saw the addition at the end about using our existing .arb file localizations on the String overrides in the custom wiredash delegate? Not sure if or how that may be possible.

**Comment by mdrideout** [comment]
> Meanwhile, please try the fix in #224 I just merged. It allows you to place `Wiredash` below `MaterialApp`

This works great. Thanks again.

**Comment by passsy** [comment]
> Can you suggest a better option for getting translated labels from our .arb files if Wiredash needs to be above MaterialApp?

You can inject the loaded `AppLocalizations` into the `_DeOverrides`. It can work like this:

```dart
class CustomWiredashTranslationsDelegate
    extends LocalizationsDelegate<WiredashLocalizations> {
  const CustomWiredashTranslationsDelegate();

  @override
  bool isSupported(Locale locale) {
    return ['de'].contains(locale.languageCode);
  }

  @override
  Future<WiredashLocalizations> load(Locale locale) async {
    if (locale.languageCode == 'de') {
      final localization = await AppLocalizations.delegate.load(locale);
      return _DeOverrides(localization);
    }
    throw "Unsupported locale $locale";
  }

  @override
  bool shouldReload(CustomWiredashTranslationsDelegate old) => false;
}
```

```dart
class _DeOverrides extends WiredashLocalizationsDe {
  final AppLocalizations localizations;

  _DeOverrides(this.localizations);

  @override
  String get feedbackStep1MessageTitle => localizations.helloWorld;
}
```

(based on the sample in https://github.com/wiredashio/wiredash-sdk/blob/stable/examples/localization/lib/main.dart)
